lib/grammar-error.js

```diff
@@ -2,10 +2,29 @@
 
 // Thrown when the grammar contains an error.
 class GrammarError extends Error {
-  constructor(message, location) {
+  constructor(message, location, referenceLocation) {
```

Comment:
I love this, and am going to make the change.  What do you think of `...locations` as the second parameter?

Reply:
You mean
```ts
constructor(message: string, locations: LocationRange[], notes: DiagnoticNote[])

type DiagnoticNote {
  message: string;
  location: LocationRange;
}
```
?
No, I'm thinking about
```ts
constructor(message: string, location?: LocationRange, notes?: DiagnoticNote[])
```

Reply:
I was thinking of just `constructor(message, ...locations)`, and locations could have diagnostic notes if you want.

Reply:
That will not cover all possible variants.

Main message should be always present. `location` could be missing, for example, check, that `allowedStartRules` contains only declared rules (pegjs/pegjs#532, I should port that stuff). `notes` available not for all diagnostics

Reply:
in that case you'd just call `new GrammarError("invalid start rule")`, right?  the message is still a separate parameter in my suggestion.

Reply:
Right. But if we merges `location` and `notes` into one single parameter, then we
- or should accept `DiagnosticNote` type, which in case when only main location is needed is slightly verbose
  ```js
  new GrammarError("message", { location });
  ```
  Besides, if we have an message without location but with some notes, it is also is ugly to represent:
  ```js
  new GrammarError("message", { location: null }, { message: "note1", location: ... });
  ```
  Of course, currently we have no such diagnostics, but who knows...
- or we should put additional diagnostic message into `LocationRange`, but that is not place for it

Reply:
I was thinking about adding an optional message to LocationRange, yes.  But I agree, that's a separate change.  Let me finish off what I have, then let's take a look at it.

Reply:
spreading a container requires that to be varargs forever.  if you decide you want to add something else afterwards, you can't.

better to just pass locations as a container imo

Reply:
Technically, you remove the `GrammarError.location` property, this is breaking change.

I thinking about this a bit. We can unify my and you approaches -- make the `GrammarError` accept just an array of `DiagnosticNote` objects. In order to keep backward compatibility I'm suggest the following:
- Currently  implement `GrammarError` as I suggested:
  ```js
  class GrammarError {
    constructor(message: string, location?: LocationRange, diagnostics?: ...DiagnosticNote) {}
  }
  // usage:
  throw new GrammarError("main message");
  throw new GrammarError("main message", main_location);
  throw new GrammarError(
    "main message", main_location,
    { message: "additional note message", location: note_location }
  );
  throw new GrammarError(
    "main message", null,
    { message: "additional note message 1", location: note_location },
    { message: "additional note message 2" }
  );
  ```
  That keeps the backward compatibility
- In version 2.0 replace this to
  ```js
  class GrammarError {
    constructor(diagnostics?: ...DiagnosticNote) {}
  }
  // usage:
  throw new GrammarError({ message: "main message" });
  throw new GrammarError({ message: "main message", location: main_location });
  throw new GrammarError(
    { message: "main message", location: main_location },
    { message: "additional note message", location: note_location }
  );
  throw new GrammarError(
    { message: "main message" },
    { message: "additional note message 1", location: note_location },
    { message: "additional note message 2" }
  );
  ```
- Release 1.2.0 in the coming month, then release 2.0 soon

Maybe even keep both variants, because passing main message and their location are common case.

Anyway, the end-user is unlikely will deal with the `GrammarError` constructor. I'll make a new PR in couple of days where I introduce the new way of reporting errors, that do not require you to throw exceptions. That will allow to emit several errors, warnings and information messages during compilation run.

What do you think?

Reply:
Yep, i just found the breaking change when I went to look at the .d.ts file.  I've added a getter that should paper over the issue, just returning the first location.  I'm willing to try implementing your API above and see how I feel about it.  Let's see what I come up with.

If you're going to not be throwing errors, is this PR all moot, or are you planning to reuse any of it in your approach?  I'd rather have multiple errors if possible.